src/hypothesis/internal/conjecture/engine.py

```diff
@@ -1187,6 +1174,98 @@ def select(self):
         return t, result
 
 
+Boring = UniqueIdentifier('Boring')
+
+
+class MultiShrinker(object):
+    """A Multishrinker takes a number of examples, a predicate, and some
+    classification function. The goal is to simultaneously find a small example
+    of each observed outcome of the classification function that satisfies the
+    predicate.
+
+    An example of this is with failing tests, where we have possibly
+    many different bugs. The predicate is that it's a failing test (i.e.
+    status is Status.INTERESTING) and the classification is the
+    interestingness_origin.
+
+    """
+
+    def __init__(self, engine, initial, predicate, classifier):
+        self.__engine = engine
+        self.__predicate = predicate
+        self.__classifier = classifier
+
+        self.__best = {}
+        self.__shrinkers = {}
+
+        self.__fully_reduced = set()
+
+        for example in initial:
+            self.classify(example)
+
+    def debug(self, msg):
+        self.__engine.debug(msg)
+
+    def classify(self, test_data):
+        if not self.__predicate(test_data):
+            return Boring
+        result = self.__classifier(test_data)
+        try:
+            existing = self.__best[result]
+            replace = sort_key(test_data.buffer) < sort_key(existing.buffer)
+        except KeyError:
+            replace = True
+        if replace:
+            self.__fully_reduced.discard(result)
+            self.__best[result] = test_data
+        return result
+
+    def shrinker(self, label):
+        current_target = self.__best[label]
+        try:
+            current_shrinker = self.__shrinkers[label]
+            if current_shrinker.shrink_target is current_target:
```

Comment:
I suspect we want equality rather than identity here; if not a comment would be appreciated.